lib/metadata/metadataUtils.js

```diff
@@ -238,6 +261,15 @@ function standardMetadataValidateBucketAndObj(params, actionImplicitDenies, log,
             }
             return next(null, bucket, objMD);
         },
+        (bucket, objMD, next) => {
+            let contentLength = request?.parsedContentLength || 0;
+            if (!contentLength && !objMD) {
+                return next(null, bucket, objMD);
+            }
+            contentLength = processBytesToWrite(request, bucket, versionId, contentLength, objMD);
+            return validateQuotas(bucket, request.accountQuotas, requestType, request.apiMethod,
+                contentLength, log, err => next(err, bucket, objMD));
```

Comment:
Putting the quota check here will impact many APIs, introducing a very slight risk (like quota unexpectedly applied on read...), as well as negigible cost (extra test), but also possibly leading to unexpected error : if an invalid request is made (like restore on a hot object), it may thus get a quota error instead of the invalid state error...

Wondering what is the expected behavior, and/or if we should change it now or in a followup ticket.

Reply:
As discussed, we will, to better centralize the logic and minimize the risks, keep the quota evaluation logic as part of the authz checks. But:

- We will handle the restore object separately, as extending the restoratin duration means no quota change.
- We will revert the inflights set in case of error at the end of an API call, if needed.

Reply:
Changes are done:

- Generalized the quota logic implementation, it can be another implementation, not always "scuba" (which is one of the possible solutions from S3 point of view). The backend is to be specifiec with `S3QUOTA`. To be also more future-proof, by default, the inflights are disabled, as this will be our nominal use case later. A configuration is thus added under `quota` to support it (ZKOP is also updated).

- To handle errors in APIs, the quota evaluation logic modifies the request object and adds some new methods that are evaluated, for each action of the API, at the end, in case of error. For example, it ensures we do not change the inflights (when enabled) if we try to delete a locked object. -**Note**: I found that we always pass the `objectReplicate` action in the route backbeat, so quotas are in fact **not** evaluated in there. This will be updated later based on the decision we make.

- The quota logic is better filtered and will only execute when matching the list of known APIs, or if we pass specific flags (e.g., during copy operation).